jans-cedarling/cedarling/src/common/cedar_schema/cedar_json/action.rs

```diff
@@ -5,27 +5,180 @@
  * Copyright (c) 2024, Gluu, Inc.
  */
 
+use crate::common::cedar_schema::cedar_json::SchemaDefinedType;
+
 use super::{
     entity_types::{
         CedarSchemaEntityAttribute, CedarSchemaEntityType, PrimitiveType, PrimitiveTypeKind,
     },
-    CedarSchemaRecord,
+    CedarSchemaEntities, CedarSchemaJson, CedarSchemaRecord, CedarType, GetCedarTypeError,
 };
 use serde::{de, ser::SerializeMap, Deserialize, Serialize};
 use serde_json::{json, Value};
 use std::collections::{HashMap, HashSet};
 
-pub type EntityRef = String;
+type AttrName = String;
+
+#[derive(Debug, Eq, Hash, PartialEq)]
+pub struct CtxAttribute {
+    pub namespace: String,
+    pub key: String,
+    pub kind: CedarType,
+}
+
+pub struct Action<'a> {
+    pub principal_entities: HashSet<String>,
+    pub resource_entities: HashSet<String>,
+    pub context_entities: HashSet<CtxAttribute>,
+    pub schema_entities: &'a CedarSchemaEntities,
+    pub schema: &'a ActionSchema,
+}
+
+impl Action<'_> {
+    /// Builds the JSON representation of context entities for a given action.
+    ///
+    /// This method processes the context attributes of the action and generates a
+    /// corresponding JSON value. The context may include entity references (with
+    /// `type` and `id`) and other values, which can be mapped through the provided
+    /// `id_mapping` and `value_mapping`.
+    ///
+    /// The `id_mapping` param is a A `HashMap` that maps context attribute keys
+    /// (like `"access_token"`) to their corresponding `id`s (like `"acs-tkn-1"`).
+    ///
+    /// # Usage Example
+    ///
+    /// ```rs
+    /// let id_mapping = HashMap::from([("access_token".to_string(), "acs-tkn-1".to_string())]);
+    /// let json = action.build_ctx_entities_json(id_mapping, value_mapping);
+    /// ```
+    pub fn build_ctx_entity_refs_json(
+        &self,
+        mut id_mapping: HashMap<String, String>,
+    ) -> Result<Value, BuildJsonCtxError> {
+        let mut json = json!({});
+
+        for attr in self.context_entities.iter() {
+            if let CedarType::TypeName(type_name) = &attr.kind {
+                let id = match id_mapping.remove(&attr.key) {
```

Comment:
Why we need remove value from HashMap? 

I don't know why, but what if User want to use same ID twice?

Reply:
anyway, using `json!({"type": type_name, "id": id}`
expands as 
```
serde_json::Value::Object({
    let mut object = serde_json::Map::new();
    let _ = object.insert(("type").into(), (serde_json::to_value(&type_name).unwrap()));
    let _ = object.insert(("id").into(), (serde_json::to_value(&id).unwrap()));
    object
})
```
so it clone string anyway, so it is OK to use `id` as &str or &String, it change nothing

Reply:
> I don't know why, but what if User want to use same ID twice?
that shouldn't be allowed. this keys in the HashMap is the keys in the cedar schema record. For example:
```cedarschema
Jans {
    action View appliesTo {
        principal: [User],
        resource: [File],
        context: {
            "status": String,
            "id_token": Id_token,  <-- lower-case key in the record
        },
    };
}
```
the keys here would be `"status"` and `"id_token"`.